src/main/java/com/es/phoneshop/dao/impl/AbstractDao.java

```diff
@@ -0,0 +1,55 @@
+package com.es.phoneshop.dao.impl;
+
+import com.es.phoneshop.dao.Dao;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+public abstract class AbstractDao<T> implements Dao<T> {
+
+    List<T> items = new ArrayList<>();
+
+    @Override
+    public synchronized void save(T item) {
+        if (item == null) {
+            throw new IllegalArgumentException("Item should not be null");
+        }
+        try {
+            get(getId(item));
+            throw new IllegalArgumentException("Item already exists");
+        } catch (NoSuchElementException e) {
+            items.add(item);
+        }
+    }
+
+    @Override
+    public synchronized T get(Long id) {
+        if (id == null) {
+            throw new IllegalArgumentException("Id should not be null");
+        }
+        return findItemById(id)
+                .orElseThrow(() -> new NoSuchElementException("Item with id " + id + " was not found"));
+    }
+
+    @Override
+    public synchronized void delete(Long id) {
+        if (id == null) {
+            throw new IllegalArgumentException("Id should not be null");
+        }
+        findItemById(id).ifPresentOrElse((item) -> {
+            items.remove(item);
+        }, () ->{
+            throw new NoSuchElementException("Item with id " + id + " was not found");
+        });
+    }
+
+    protected abstract Long getId(T t);
```

Comment:
What for? Make use of java bounded type parameters in daos not to implement workarounds like this